hepcrawl/spiders/edp_spider.py

```diff
@@ -0,0 +1,430 @@
+
+from ..extractors.jats import Jats
+from ..items import HEPRecord
+from ..loaders import HEPLoader
+from ..utils import (
+    ftp_list_files,
+    ftp_connection_info,
+    get_first,
+    get_node,
+    parse_domain,
+)
+
+
+class EDPSpider(Jats, XMLFeedSpider):
+    """EDP Sciences crawler.
+
+    This spider connects to a given FTP hosts and downloads zip files with
+    XML files for extraction into HEP records.
+
+    This means that it generates the URLs for Scrapy to crawl in a special way:
+
+    1. First it connects to a FTP host and lists all the new TAR files found
+       on the remote server and downloads them to a designated local folder,
+       using `start_requests()`. The starting point of the crawl can also be a
+       local file. Packages contain XML files with different formats (gz package
+       is JATS, bz2 package has "rich" and "jp" format XML files, "jp" is JATS.)
+
+    2. Then the TAR file is unpacked and it lists all the XML files found
+       inside, via `handle_package()`. Note the callback from `start_requests()`
+
+    3. Each XML file is parsed via `parse_node()`.
+
+    4. PDF file is fetched from the web page in `scrape_for_pdf()`.
+
+    5. Finally, HEPRecord is created in `build_item()`.
+
+    To run a crawl, you need to pass FTP connection information via
+    `ftp_host` and `ftp_netrc`:``
+
+    .. code-block:: console
+
+        scrapy crawl EDP -a ftp_host=ftp.edpsciences.org -a ftp_netrc=tmp/edps_netrc
+        scrapy crawl EDP -a package_path=file://`pwd`/tests/responses/edp/test_jp.tar.bz2
+
+    Happy crawling!
+    """
+
+    # FIXME: Sometimes there are errors:
+    # Unhandled Error
+    # self.f.seek(-size-self.SIZE_SIZE, os.SEEK_END)
+    # exceptions.IOError: [Errno 22] Invalid argument
+    #
+    # OR:
+    # ConnectionLost: ('FTP connection lost',
+    # <twisted.python.failure.Failure twisted.internet.error.ConnectionDone:
+    # Connection was closed cleanly.>)
+
+    # NOTE: See old harvesting-kit:
+    # https://github.com/inspirehep/harvesting-kit/blob/master/harvestingkit/edpsciences_package.py
+    # https://github.com/inspirehep/inspire/blob/master/bibtasklets/bst_edpsciences_harvest.py
+
+    name = 'EDP'
+    custom_settings = {}
+    start_urls = []
+    iterator = 'xml'
+    itertag = 'article'
+
+    allowed_article_types = [
+        'research-article',
+        'corrected-article',
+        'original-article',
+        'introduction',
+        'letter',
+        'correction',
+        'addendum',
+        'review-article',
+        'rapid-communications',
+        'Article',
+        'Erratum',
+    ]
+
+    def __init__(self, package_path=None, ftp_folder="incoming", ftp_host=None, ftp_netrc=None, *args, **kwargs):
+        """Construct EDP spider."""
+        super(EDPSpider, self).__init__(*args, **kwargs)
+        self.ftp_folder = ftp_folder
+        self.ftp_host = ftp_host
+        self.ftp_netrc = ftp_netrc
+        self.target_folder = "/tmp/EDP"
```

Comment:
This might be better using a config value or equivalent, a good point for future improvement. Maybe using the mktmp if it's only used by this process.